### PR TITLE
Push function pointer as CFunction

### LINKED ISSUE #46: [ENH] Add C# function pointers support for ILua.PushCFunction
# Description

We should make use of C# 9.0 [function pointers](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/function-pointers.md) and [UnmanagedCallersOnly attribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute?view=net-5.0) for better performance and code maintenance.

# TODO

- [ ] Add `PushCFunction(delegate* unmanaged[Cdecl]<IntPtr, int>)` method definition to `ILua` interface.

- [ ] Add method implementation to to `Lua` class.

- [ ] Add corresponding test.

### PR DESCRIPTION
**Pull request by GlebChili**

### CODE REVIEW

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        public Task<bool> Start(ILua lua, GetILuaFromLuaStatePointer lua_extructor, ModuleAssemblyLoadContext assembly_context)
+        {
+            TaskCompletionSource<bool> taskCompletion = new();
+
+            try
+            {
+                unsafe
+                {
+                    lua.PushCFunction((delegate* unmanaged[Cdecl]<IntPtr, int>)(delegate* managed<IntPtr, int>)&TestFunc);
```
Comment (GlebChili) [inline gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: This horrible double type cast will go away as soon as .NET 5 SDK RC2 is available.

Comment (AaronJThompson) [reply gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: Shall we wait for RC2 before merging this?

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                    lua.Pop(2);
+
+                    if(ret_string != random_string)
+                    {
+                        throw new Exception("Returned string is invalid");
+                    }
+
+                    if(ret_number != random_number)
+                    {
+                        throw new Exception("Returned number is invalid");
```
Comment (AaronJThompson) [inline gm_dotnet_managed/Tests/PushCFuncPtr.cs:61]: Just to confirm, is the standard styling for the project always braces even for single line bodies?

File: gm_dotnet_managed/GmodNET/Startup.cs [hunk 3]
```diff
@@ -156,7 +156,10 @@ internal static IntPtr Main(IntPtr lua_base, IntPtr native_version_string, int v
                     LuaInterop.check_number = CreateNativeCaller<Func<IntPtr, int, double>>(params_from_native_code[53]);
 
                     LuaInterop.push_c_function_safe = CreateNativeCaller<Action<IntPtr, IntPtr, IntPtr>>(params_from_native_code[54]);
+
+                    GmodInterop.lua_extructor = &LuaInterop.ExtructLua;
```
Comment (AaronJThompson) [inline gm_dotnet_managed/GmodNET/Startup.cs]: Wait, do you mean Extract rather than Extruct?

File: gm_dotnet_managed/GmodNET/Lua.cs [hunk 4]
```diff
@@ -367,11 +368,28 @@ public void PushCFunction(CFuncManagedDelegate managed_function, bool use_safe_e
...
+        public unsafe void PushCFunction(delegate* unmanaged[Cdecl]<IntPtr, int> function_pointer)
+        {
+            IntPtr int_ptr = (IntPtr)function_pointer;
+
+            if (int_ptr == IntPtr.Zero)
+            {
+                throw new ArgumentNullException("function_pointer", "Parameter can't be nullptr.");
+            }
+
+            push_c_function(ptr, (IntPtr)function_pointer);
```
Comment (AaronJThompson) [inline gm_dotnet_managed/GmodNET/Lua.cs]: Use local var int_ptr instead of casting again

**Review by AaronJThompson — CHANGES_REQUESTED** [review]

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs (hunk 1 above)
Comment (GlebChili) [reply gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: Will fix as soon as RC2 is out, works for now so we can merge as it is.

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs (hunk 2 above)
Comment (GlebChili) [reply gm_dotnet_managed/Tests/PushCFuncPtr.cs:61]: Kind of. I always write like this, so all other parts of project are formatted in the same manner. So yes, we can consider it as a part of project's code style.

File: gm_dotnet_managed/GmodNET/Startup.cs (hunk 3 above)
Comment (GlebChili) [reply gm_dotnet_managed/GmodNET/Startup.cs]: Oops, will fix it now 😂

File: gm_dotnet_managed/Tests/PushCFuncPtr.cs (hunk 1 above)
Comment (GlebChili) [reply gm_dotnet_managed/Tests/PushCFuncPtr.cs:46]: It is a test, so there will be no changes to the core codebase anyway.

File: gm_dotnet_managed/GmodNET/Lua.cs (hunk 4 above)
Comment (GlebChili) [reply gm_dotnet_managed/GmodNET/Lua.cs]: Fixed

**Comment by GlebChili** [comment]
@AaronJThompson Take a look again please.

**Review by AaronJThompson — APPROVED** [review]
Looks good. Approved.
As a side note, it would be awesome if we could add the XMLCommetns to the ILua.TYPES enum values to avoid the warnings and provide context for developers

**Pull request merged**